README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
BuenColors
==========

A package that facilitates nice colors created for the [Buenrostro Lab](https://buenrostrolab.com). The base of the package including some divergent color schemes were taken from the [wesanderson](https://github.com/karthik/wesanderson) package. This README provides a quick overview of how to use the color schemes with ggplot2 and which palettes are available.

Installation
------------

``` r
devtools::install_github("caleblareau/BuenColors")
```

Color Palettes
--------------

``` r
library(BuenColors)
```

With the library loaded, just type this to get either the `continuous` or `discrete` (by default) palette printed in your `plot` console.

``` r
jdb_palette("FantasticFox")
```

![](figure/example-1.png)

``` r
jdb_palette("aqua_brick", type = "continuous")
```

![](figure/example-2.png)

### Here are all the names that are available...

    #>   [1] "algae_earth"    "aqua_brick"     "aqua_tan"       "berry"         
    #>   [5] "blue_cyan"      "BottleRocket"   "BottleRocket2"  "brewer_blue"   
    #>   [9] "brewer_celsius" "brewer_fire"    "brewer_green"   "brewer_heat"   
    #>  [13] "brewer_jamaica" "brewer_marine"  "brewer_orange"  "brewer_purple" 
    #>  [17] "brewer_red"     "brewer_spectra" "brewer_violet"  "brewer_yes"    
    #>  [21] "calma_azules"   "calma_bosque"   "calma_manudo"   "calma_marino"  
    #>  [25] "calma_morado"   "calma_musgos"   "Cavalcanti"     "Chevalier"     
    #>  [29] "china_basics"   "china_dragon"   "china_novice"   "china_ranges"  
    #>  [33] "china_sunset"   "china_theory"   "china_weirdo"   "citric"        
    #>  [37] "citric_yellow"  "citrus"         "cyan_brick"     "cyan_green"    
    #>  [41] "cyan_pink"      "cyan_purple"    "cyan_tan"       "cyan_violet"   
    #>  [45] "Darjeeling"     "Darjeeling2"    "dark_blue"      "dark_citrus"   
    #>  [49] "dark_cyan"      "dark_violet"    "dusk_dawn"      "FantasticFox"  
    #>  [53] "flame_artic"    "flame_blind"    "flame_flame"    "flame_light"   
    #>  [57] "flame_macaw"    "flame_polar"    "flame_volts"    "flame_watts"   
    #>  [61] "flame_weird"    "flame_wings"    "forest"         "forest_citric" 
    #>  [65] "forest_yellow"  "GrandBudapest"  "GrandBudapest2" "horizon"       
    #>  [69] "horizon_extra"  "lawhoops"       "Moonrise1"      "Moonrise2"     
    #>  [73] "Moonrise3"      "ocean_aqua"     "ocean_brick"    "ocean_citrus"  
    #>  [77] "ocean_earth"    "ocean_green"    "ocean_pink"     "ocean_red"     
    #>  [81] "ocean_teal"     "purple_baby"    "purple_pink"    "Royal1"        
    #>  [85] "Royal2"         "Rushmore"       "samba_color"    "samba_light"   
    #>  [89] "samba_night"    "solar_basic"    "solar_blues"    "solar_extra"   
    #>  [93] "solar_flare"    "solar_glare"    "solar_rojos"    "teal_orange"   
    #>  [97] "teal_violet"    "white_grove"    "white_jungle"   "white_mango"   
    #> [101] "white_orange"   "white_tango"    "wolfgang_basic" "wolfgang_extra"
    #> [105] "Zissou"

Color Maps
----------

To keep consistent color designations, one can use the `color_map` function to link features to their specific hex color annoations. For example,

``` r
jdb_color_map(c("HSC"))
#> [1] "#00441B"
```

returns the hex code associated with `HSC` in the Buenrostro Lab paradigm. This function may be applied over multiple features--

``` r
jdb_color_map(c("HSC", "CMP", "HSC"))
#> [1] "#00441B" "#FFC179" "#00441B"
```

and will error out when a feature is not recognized--

``` r
jdb_color_map(c("WHAT"))
#>  Error: all(name %in% names(jdb_color_maps)) is not TRUE 
```

### Here are all the names that are available...

    #>  [1] "B"     "CD4"   "CD8"   "CLP"   "CMP"   "Ery"   "GMP"   "GMP-A"
    #>  [9] "GMP-B" "GMP-C" "HSC"   "LMPP"  "MEP"   "mono"  "MPP"   "NK"   
    #> [17] "pDC"

### Here are what the mappings look like...

![](figure/colormaps-1.png)

ggplot example
--------------

To coordinate a ggplot feature (e.g. data point in a scatter plot) with a particular color, [this post](https://www.biostars.org/p/204891/) was a life-saver. Specifically, we'll use a named vector to coordinate the discrete values. Here's an example--

``` r
xy <- 1:7
cell <- c("GMP-A", "Ery", "CD4", "Ery", "LMPP", "ERY", "MEP")

df <- data.frame(
  xy = xy,
  cell = cell, stringsAsFactors = FALSE
)

ggplot(df, aes(x = xy, y = xy, color = cell)) +
  geom_point(size = 10) + pretty_plot() +
  scale_color_manual(values = jdb_color_maps)
#> Warning: Removed 1 rows containing missing values (geom_point).
```

![](figure/colormapPlotExample-1.png)

**P.S.**-- this call will return a blank color for features that are not found in the color map ("ERY" in this example).

### Important Note on color mappings...

The above `ggplot` command works because `jdb_color_maps` (with an 's') exists as a named vector in the `BuenColors` NAMESPACE. The provided function (`jdb_color_map`) does not have an 's' by the way. This same syntax of supplying a named vector should work for all discrete color scale functionalities in `ggplot`.

Discrete colors
===============

The trick here is to use `scale_color_manual` like it is shown here--

``` r
library(ggplot2)
ggplot(iris, aes(Sepal.Length, Sepal.Width, color = Species)) + 
  geom_point() + pretty_plot() + 
  scale_color_manual(values = jdb_palette("brewer_spectra"))
```

![](figure/discrete-1.png)

Continuous fill
===============

The trick here is to use `scale_color_gradientn` like it is shown here--

``` r
df <- data.frame(x = rnorm(1000), y = 0)
ggplot(df, aes(x=x, y=y, colour=x)) + geom_point() + 
  scale_color_gradientn(colors = jdb_palette("flame_light")) +
  pretty_plot()
```

![](figure/ggplotcont-1.png)

Density Plot
============

Best way that I've found to make the density color function represented in the points. Thanks to [Kamil Slowikowski](http://slowkow.com/notes/ggplot2-color-by-density/) for figuring this out.

``` r
dat <- data.frame(
  x = c(
    rnorm(1e4, mean = 0, sd = 0.1),
    rnorm(1e3, mean = 0, sd = 0.1)
  ),
  y = c(
    rnorm(1e4, mean = 0, sd = 0.1),
    rnorm(1e3, mean = 0.1, sd = 0.2)
  )
)
dat$density <- get_density(dat$x, dat$y)
ggplot2::ggplot(dat) + geom_point(aes(x, y, color = density)) + 
  scale_color_gradientn(colors = jdb_palette("solar_extra")) +
  pretty_plot()
```

![](figure/ggplotdensty-1.png)

Shuffle Plot Order
==================

Quick wrapper using `shuf` to change the order of plotting points (to a random presentation) to avoid hiding effects.

``` r
tdf<-paste(system.file('rds',package='BuenColors'),'basicTSNE.rds',sep='/')
df <- readRDS(tdf)
ggplot(shuf(df)) + geom_point(aes(X1, X2, color = counts)) + 
  scale_color_gradientn(colors = jdb_palette("solar_extra")) +
  pretty_plot()
```

![](figure/tsne-1.png)

Continuous Colors
-----------------

Here's what each palette looks like on a continuous scale.

![](figure/printCont-1.png)![](figure/printCont-2.png)![](figure/printCont-3.png)![](figure/printCont-4.png)![](figure/printCont-5.png)![](figure/printCont-6.png)![](figure/printCont-7.png)![](figure/printCont-8.png)![](figure/printCont-9.png)![](figure/printCont-10.png)![](figure/printCont-11.png)![](figure/printCont-12.png)![](figure/printCont-13.png)![](figure/printCont-14.png)![](figure/printCont-15.png)![](figure/printCont-16.png)![](figure/printCont-17.png)![](figure/printCont-18.png)![](figure/printCont-19.png)![](figure/printCont-20.png)![](figure/printCont-21.png)![](figure/printCont-22.png)![](figure/printCont-23.png)![](figure/printCont-24.png)![](figure/printCont-25.png)![](figure/printCont-26.png)![](figure/printCont-27.png)![](figure/printCont-28.png)![](figure/printCont-29.png)![](figure/printCont-30.png)![](figure/printCont-31.png)![](figure/printCont-32.png)![](figure/printCont-33.png)![](figure/printCont-34.png)![](figure/printCont-35.png)![](figure/printCont-36.png)![](figure/printCont-37.png)![](figure/printCont-38.png)![](figure/printCont-39.png)![](figure/printCont-40.png)![](figure/printCont-41.png)![](figure/printCont-42.png)![](figure/printCont-43.png)![](figure/printCont-44.png)![](figure/printCont-45.png)![](figure/printCont-46.png)![](figure/printCont-47.png)![](figure/printCont-48.png)![](figure/printCont-49.png)![](figure/printCont-50.png)![](figure/printCont-51.png)![](figure/printCont-52.png)![](figure/printCont-53.png)![](figure/printCont-54.png)![](figure/printCont-55.png)![](figure/printCont-56.png)![](figure/printCont-57.png)![](figure/printCont-58.png)![](figure/printCont-59.png)![](figure/printCont-60.png)![](figure/printCont-61.png)![](figure/printCont-62.png)![](figure/printCont-63.png)![](figure/printCont-64.png)![](figure/printCont-65.png)![](figure/printCont-66.png)![](figure/printCont-67.png)![](figure/printCont-68.png)![](figure/printCont-69.png)![](figure/printCont-70.png)![](figure/printCont-71.png)![](figure/printCont-72.png)![](figure/printCont-73.png)![](figure/printCont-74.png)![](figure/printCont-75.png)![](figure/printCont-76.png)![](figure/printCont-77.png)![](figure/printCont-78.png)![](figure/printCont-79.png)![](figure/printCont-80.png)![](figure/printCont-81.png)![](figure/printCont-82.png)![](figure/printCont-83.png)![](figure/printCont-84.png)![](figure/printCont-85.png)![](figure/printCont-86.png)![](figure/printCont-87.png)![](figure/printCont-88.png)![](figure/printCont-89.png)![](figure/printCont-90.png)![](figure/printCont-91.png)![](figure/printCont-92.png)![](figure/printCont-93.png)![](figure/printCont-94.png)![](figure/printCont-95.png)![](figure/printCont-96.png)![](figure/printCont-97.png)![](figure/printCont-98.png)![](figure/printCont-99.png)![](figure/printCont-100.png)![](figure/printCont-101.png)![](figure/printCont-102.png)![](figure/printCont-103.png)![](figure/printCont-104.png)![](figure/printCont-105.png)

Discrete colors
---------------

Here are the discrete color units that go into each scale. If n is small and discrete, note that the colors that are selected are read from left to right.

![](figure/printDiscrete-1.png)![](figure/printDiscrete-2.png)![](figure/printDiscrete-3.png)![](figure/printDiscrete-4.png)![](figure/printDiscrete-5.png)![](figure/printDiscrete-6.png)![](figure/printDiscrete-7.png)![](figure/printDiscrete-8.png)![](figure/printDiscrete-9.png)![](figure/printDiscrete-10.png)![](figure/printDiscrete-11.png)![](figure/printDiscrete-12.png)![](figure/printDiscrete-13.png)![](figure/printDiscrete-14.png)![](figure/printDiscrete-15.png)![](figure/printDiscrete-16.png)![](figure/printDiscrete-17.png)![](figure/printDiscrete-18.png)![](figure/printDiscrete-19.png)![](figure/printDiscrete-20.png)![](figure/printDiscrete-21.png)![](figure/printDiscrete-22.png)![](figure/printDiscrete-23.png)![](figure/printDiscrete-24.png)![](figure/printDiscrete-25.png)![](figure/printDiscrete-26.png)![](figure/printDiscrete-27.png)![](figure/printDiscrete-28.png)![](figure/printDiscrete-29.png)![](figure/printDiscrete-30.png)![](figure/printDiscrete-31.png)![](figure/printDiscrete-32.png)![](figure/printDiscrete-33.png)![](figure/printDiscrete-34.png)![](figure/printDiscrete-35.png)![](figure/printDiscrete-36.png)![](figure/printDiscrete-37.png)![](figure/printDiscrete-38.png)![](figure/printDiscrete-39.png)![](figure/printDiscrete-40.png)![](figure/printDiscrete-41.png)![](figure/printDiscrete-42.png)![](figure/printDiscrete-43.png)![](figure/printDiscrete-44.png)![](figure/printDiscrete-45.png)![](figure/printDiscrete-46.png)![](figure/printDiscrete-47.png)![](figure/printDiscrete-48.png)![](figure/printDiscrete-49.png)![](figure/printDiscrete-50.png)![](figure/printDiscrete-51.png)![](figure/printDiscrete-52.png)![](figure/printDiscrete-53.png)![](figure/printDiscrete-54.png)![](figure/printDiscrete-55.png)![](figure/printDiscrete-56.png)![](figure/printDiscrete-57.png)![](figure/printDiscrete-58.png)![](figure/printDiscrete-59.png)![](figure/printDiscrete-60.png)![](figure/printDiscrete-61.png)![](figure/printDiscrete-62.png)![](figure/printDiscrete-63.png)![](figure/printDiscrete-64.png)![](figure/printDiscrete-65.png)![](figure/printDiscrete-66.png)![](figure/printDiscrete-67.png)![](figure/printDiscrete-68.png)![](figure/printDiscrete-69.png)![](figure/printDiscrete-70.png)![](figure/printDiscrete-71.png)![](figure/printDiscrete-72.png)![](figure/printDiscrete-73.png)![](figure/printDiscrete-74.png)![](figure/printDiscrete-75.png)![](figure/printDiscrete-76.png)![](figure/printDiscrete-77.png)![](figure/printDiscrete-78.png)![](figure/printDiscrete-79.png)![](figure/printDiscrete-80.png)![](figure/printDiscrete-81.png)![](figure/printDiscrete-82.png)![](figure/printDiscrete-83.png)![](figure/printDiscrete-84.png)![](figure/printDiscrete-85.png)![](figure/printDiscrete-86.png)![](figure/printDiscrete-87.png)![](figure/printDiscrete-88.png)![](figure/printDiscrete-89.png)![](figure/printDiscrete-90.png)![](figure/printDiscrete-91.png)![](figure/printDiscrete-92.png)![](figure/printDiscrete-93.png)![](figure/printDiscrete-94.png)![](figure/printDiscrete-95.png)![](figure/printDiscrete-96.png)![](figure/printDiscrete-97.png)![](figure/printDiscrete-98.png)![](figure/printDiscrete-99.png)![](figure/printDiscrete-100.png)![](figure/printDiscrete-101.png)![](figure/printDiscrete-102.png)![](figure/printDiscrete-103.png)![](figure/printDiscrete-104.png)![](figure/printDiscrete-105.png)


NEWS

Version 0.5.4
=============
* Parameterize font in pretty_plot()

Version 0.5.3
=============
* Arial becomes default font in pretty_plot()
* Added g_legend() function

Version 0.5.1
=============
* Minor color map tweaks

Version 0.5.0
=============
* Initial submission to CRAN

Version 0.4.3
=============
* Better Handling of NAs

Version 0.4.2
=============
* Enables mapping of continuous vector to continuous color palette

Version 0.4.0
=============
* More custom color palettes

Version 0.3.1
=============
* Remove grey boxes automatically when doing facets.

Version 0.3.0
=============
* Celltype/Feature color maping.

Version 0.2.1
=============
* Starting with custom palettes for CAL.

Version 0.2
=============
* Added functionality, including pretty_plot().

Version 0.1
=============
* Basic setup.


R/plotHelp.R
#' @include colors.R
NULL

#' Compute the density of points in a grid
#'
#' Credit to Kamil Slowikowski
#' See post: http://slowkow.com/notes/ggplot2-color-by-density/
#'
#' @param x X coordinates
#' @param y Y coordinates
#' @param n Number of bins to divide grid
#' @return A vector of densities for plotting
#' @importFrom MASS kde2d
#' @keywords point density
#' @examples
#' dat <- data.frame(
#'   x = c(
#'     rnorm(1e4, mean = 0, sd = 0.1),
#'     rnorm(1e3, mean = 0, sd = 0.1)
#'   ),
#'   y = c(
#'     rnorm(1e4, mean = 0, sd = 0.1),
#'     rnorm(1e3, mean = 0.1, sd = 0.2)
#'   )
#' )
#' dat$density <- get_density(dat$x, dat$y)
#' ggplot2::ggplot(dat) + geom_point(aes(x, y, color = density))
#' @export
get_density <- function(x, y, n = 200) {
  dens <- MASS::kde2d(x = x, y = y, n = n)
  ix <- findInterval(x, dens$x)
  iy <- findInterval(y, dens$y)
  ii <- cbind(ix, iy)
  return(dens$z[ii])
}

#' Get rid of gridlines, etc.
#'
#' Idea is to make it look more like Matlab
#'
#' @param fontsize You know, the font size lol; default = 10
#' @param font The font; Arial comes non-default apparently; default = "Helvetica"
#' @return A vector of densities for plotting
#' @importFrom ggplot2 theme theme_bw element_blank element_text element_rect
#' @keywords point density
#' @examples
#' df <- data.frame(x = rnorm(1000), y = 0)
#' ggplot(df, aes(x=x, y=y, colour=x)) + geom_point()
#'
#' @export
pretty_plot <- function(fontsize = 10, font = "Helvetica"){
nl <- theme_bw(base_size = fontsize) +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    axis.text = element_text(colour = "black", family=font),
    legend.key = element_blank(),
    strip.background = element_rect(colour="black", fill = "white")
  )
return(nl)
}

#' Randomly shuffle the plotting order
#'
#' Idea is to make the plot independent of the sample ordering
#'
#' @param df data.frame (or similar 2D feature) whose rows will be shuffled
#' @return An object with the rows shuffled
#' @keywords point density
#' @examples
#' df <- data.frame(x = rnorm(1000), y = 0)
#' ggplot(shuf(df), aes(x=x, y=y, colour=x)) + geom_point()
#'
#' @export
shuf <- function(df){
  return(df[sample(1:dim(df)[1], dim(df)[1]),])
}

#' Maps numeric vector to color palette
#'
#' Modified from Sushi::maptocolors
#'
#' @param numberVec Numeric vector
#'
#' @param palette color palette to which to be mapped. Any
#' valid palette in "BuenColors" should work.
#'
#' @param range The c(min, max) numeric values to
#' make the color at the extremes.
#'
#' @importFrom grDevices colorRampPalette
#'
#' @examples
#' i <- rnorm(1000)
#' df <- data.frame(x = i, y = 0, color = numberToColorVec(i^2, "brewer_heat"))
#' ggplot(shuf(df), aes(x=x, y=y, colour= color)) + geom_point() +
#'  scale_color_manual(values = levels(df$color))
#'
#' @export
#'
numberToColorVec <- function(numberVec, palette, range = NULL){

  stopifnot(is.numeric(numberVec))

  # Parse numeric vector and setup breaks being sensitive to
  # user-defined caps in range
  vec <- numberVec
  num <- 100

  if (is.null(range)) {
    breaks <- seq(min(vec, na.rm = TRUE), max(vec, na.rm = TRUE), length.out = num)
  } else {
    vec[which(vec < range[1])] <- range[1]
    vec[which(vec > range[2])] <- range[2]
    breaks <- seq(range[1], range[2], length.out = num)
  }

  # Setup color vector
  col <- grDevices::colorRampPalette(jdb_palette(palette, length(jdb_palettes[[palette]]), "continuous"))
  cols <- col(length(breaks) + 1)
  colvec <- as.character(cut(vec, c(-Inf, breaks, Inf), labels=cols))
  return(colvec)
}

#' Extract a legend from ggplot
#'
#' Grab the legend from a ggplot object and use it
#' for saving in its own file
#'
#' @param a.gplot ggplot plot object with a color / fill legend
#' @return A new ggplot object with only the legend
#' @importFrom ggplot2 ggplot_gtable
#' @importFrom ggplot2 ggplot_build
#' @keywords legend guide
#' @examples
#' df <- data.frame(x = rnorm(1000), y = 0)
#' p1 <- ggplot(shuf(df), aes(x=x, y=y, colour=x)) + geom_point()
#' legendP <- g_legend(p1)
#'
#' @export
# Extract legend
g_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)
}




R/cell_colors.R
#' @include plotHelp.R
NULL

#' List of maps between annotations and specified colors.
#'
#' Use \code{\link{jdb_color_map}} to map cell/feature to color hex.
#'
#' @export
jdb_color_maps <- c(
  "HSC" = "#00441B",
  "MPP" = "#46A040",
  "LMPP" = "#00AF99",
  "CMP" = "#FFC179",
  "CLP" = "#98D9E9",
  "MEP" = "#F6313E",
  "GMP"= "#FFA300",
  "pDC" = "#C390D4",
  "mono" = "#FF5A00",
  "GMP-A" = "#AFAFAF",
  "GMP-B" = "#7D7D7D",
  "GMP-C" =  "#4B4B4B",
  "Ery" = "#8F1336",
  "CD4" = "#0081C9",
  "CD8" = "#001588",
  "NK" = "#490C65",
  "B" = "#BA7FD0"
)

#' A mapping function to specific colors.
#'
#' This function maps a common feature/cell
#' type (e.g. "HSC") and returns the canonical
#' hex code associtated with it.
#'
#' color_map (no 's') is the function; color_maps
#' is a named character vector.
#'
#' @param name a character (vector) corresponding to a
#' name with a mappable color
#' @return a character (vector) of hex codes
#' @keywords colors
#' @examples
#' # See possible values
#' names(jdb_color_maps)
#'
#' # Get hex codes
#' jdb_color_map("HSC")
#' jdb_color_map(c("HSC", "LMPP", "HSC"))
#'
#' @export
jdb_color_map <- function(name) {
  stopifnot(all(name %in% names(jdb_color_maps)))
return(unname(jdb_color_maps[name]))
}


#' List of maps between annotations and specified colors for scHeme/UKBB
#' project
#'
#' Use ejc_color_map to map cell/feature to color hex.
#'
#' @export
ejc_color_maps <- c(
  "HSC" = "#00441B",
  "MPP" = "#46A040",
  "LMPP" = "#00AF99",
  "CMP" = "#FFC179",
  "CLP" = "#98D9E9",
  "MEP" = "#F6313E",
  "pDC" = "#CDA2DB",
  "Mono" = "#FF5A00",
  "GMP-A" = "#FFCE00",
  "GMP-B" = "#FFA300",
  "GMP-C" =  "#FF7900",
  "GMP" = "#FFA300",
  "Ery" = "#8F1336",
  "CD4" = "#0081C9",
  "CD8" = "#001588",
  "NK" = "#490C65",
  "B" = "#A65AC2",
  "Mega" = "#FF81AF",
  "mDC"= "#FFEC00"
)

#' A mapping function to specific colors for UKBB
#' single-cell heme paper.
#'
#' This function maps a common feature/cell
#' type (e.g. "HSC") and returns the canonical
#' hex code associtated with it.
#'
#' color_map (no 's') is the function; color_maps
#' is a named character vector.
#'
#' @param name a character (vector) corresponding to a
#' name with a mappable color
#' @return a character (vector) of hex codes
#' @keywords colors
#' @examples
#' # See possible values
#' names(ejc_color_maps)
#'
#' # Get hex codes
#' ejc_color_map("HSC")
#' ejc_color_map(c("HSC", "LMPP", "HSC"))
#'
#' @export
ejc_color_map <- function(name) {
  stopifnot(all(name %in% names(ejc_color_maps)))
return(unname(ejc_color_maps[name]))
}

#' List of maps between annotations and specified colors
#' for the TF1 phylogeny analyses.
#'
#' Use tf1_color_map to map cell/feature to color hex.
#'
#' @export
tf1_color_maps <- c(
  "C9" = "#F6313E",
  "D6" = "#46A040",
  "G10" = "#0081C9",

  "G11" = "#A65AC2", # 21
  "B3" = "#FFA300", # 13
  "D2" = "#FFFF32", # 5
  "B5" = "#89774A", #5
  "B11" = "#FF6A80", #5
  "F4" = "#999999", #5
  "A9" = "#0DB2AA", #3

  "bulk" = "#001588",
  "mix"= "#00441B",

  "B9" = "#CDA2DB",
  "C10" = "#98D9E9",
  "C4" = "#FF9999",
  "C7" =  "#FFC966",
  "D3" = "#A9FBA9"
)

#' A mapping function to specific colors for the
#' TF1 clone paper
#'
#' This function maps a common feature/cell
#' type (e.g. "B9") and returns the canonical
#' hex code associtated with it.
#'
#' color_map (no 's') is the function; color_maps
#' is a named character vector.
#'
#' @param name a character (vector) corresponding to a
#' name with a mappable color
#' @return a character (vector) of hex codes
#' @keywords colors
#' @examples
#' # See possible values
#' names(tf1_color_maps)
#'
#' # Get hex codes
#' tf1_color_map("B9")
#' tf1_color_map(c("C9", "D6", "G10"))
#'
#' @export
tf1_color_map <- function(name) {
  stopifnot(all(name %in% names(tf1_color_maps)))
return(unname(tf1_color_maps[name]))
}


